src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -815,6 +874,131 @@ NodeListInsertCommand(List *workerNodeList)
 }
 
 
+/*
+ * MarkObjectDistributedCreateCommand generates a command that can be executed to
+ * insert or update the provided object into pg_dist_object on a worker node.
+ */
+char *
+MarkObjectDistributedCreateCommand(const ObjectAddress *address,
+								   int32 *distributionArgumentIndex,
+								   int32 *colocationId)
+{
+	StringInfo insertDistributedObjectCommand = makeStringInfo();
+	List *names;
+	List *args;
+	char *objectType;
+#if PG_VERSION_NUM >= PG_VERSION_14
+	objectType = getObjectTypeDescription(address, false);
+	getObjectIdentityParts(address, &names, &args, false);
+#else
+	objectType = getObjectTypeDescription(address);
+	getObjectIdentityParts(address, &names, &args);
+#endif
+
+	appendStringInfo(insertDistributedObjectCommand,
+					 "SELECT citus_internal_add_object_metadata(");
+
+	appendStringInfo(insertDistributedObjectCommand,
+					 "%s, ARRAY[",
+					 quote_literal_cstr(objectType));
+
+	char *name;
+	bool firstLoop = true;
+	foreach_ptr(name, names)
+	{
+		if (!firstLoop)
+		{
+			appendStringInfo(insertDistributedObjectCommand, ", ");
+		}
+		firstLoop = false;
+		appendStringInfoString(insertDistributedObjectCommand, quote_literal_cstr(name));
+	}
+
+	appendStringInfo(insertDistributedObjectCommand, "]::text[], ARRAY[");
+
+	char *arg;
+	firstLoop = true;
+	foreach_ptr(arg, args)
+	{
+		if (!firstLoop)
+		{
+			appendStringInfo(insertDistributedObjectCommand, ", ");
+		}
+		firstLoop = false;
+		appendStringInfoString(insertDistributedObjectCommand, quote_literal_cstr(arg));
+	}
+
+	appendStringInfo(insertDistributedObjectCommand, "]::text[],");
+
+	if (distributionArgumentIndex == NULL)
+	{
+		appendStringInfo(insertDistributedObjectCommand, "NULL, ");
+	}
+	else
+	{
+		appendStringInfo(insertDistributedObjectCommand, "%d, ",
+						 *distributionArgumentIndex);
+	}
+	if (colocationId == NULL)
+	{
+		appendStringInfo(insertDistributedObjectCommand, "NULL)");
+	}
+	else
+	{
+		appendStringInfo(insertDistributedObjectCommand, "%d)",
+						 *colocationId);
+	}
+
+	return insertDistributedObjectCommand->data;
+}
+
+
+/*
+ * citus_internal_add_object_metadata is an internal UDF to
+ * add a row to pg_dist_object.
+ */
+Datum
+citus_internal_add_object_metadata(PG_FUNCTION_ARGS)
+{
+	char *textType = TextDatumGetCString(PG_GETARG_DATUM(0));
+	ArrayType *nameArray = PG_GETARG_ARRAYTYPE_P(1);
+	ArrayType *argsArray = PG_GETARG_ARRAYTYPE_P(2);
+	int distributionArgumentIndexValue;
+	int colocationIdValue;
+	int *colocationId = NULL;
+	int *distributionArgumentIndex = NULL;
+
+	if (!PG_ARGISNULL(3))
+	{
+		distributionArgumentIndexValue = PG_GETARG_INT32(3);
+		distributionArgumentIndex = &distributionArgumentIndexValue;
+	}
+	if (!PG_ARGISNULL(4))
+	{
+		colocationIdValue = PG_GETARG_INT32(4);
+		colocationId = &colocationIdValue;
```

Comment:
Ah, we don't sync pg_dist_colocation, so the colocation information on the workers would not be useful?

Reply:
I think syncing pg_dist_colocation would be very useful to enable create_distributed_table from workers, but colocationid refers to pg_dist_partition, not pg_dist_colocation